perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_group_attribute_def_virt_adDisplayName_o365mu.java

```diff
@@ -0,0 +1,166 @@
+import cz.metacentrum.perun.core.api.exceptions.WrongAttributeAssignmentException;
+import cz.metacentrum.perun.core.impl.PerunSessionImpl;
+import cz.metacentrum.perun.core.implApi.modules.attributes.GroupVirtualAttributesModuleAbstract;
+import cz.metacentrum.perun.core.implApi.modules.attributes.GroupVirtualAttributesModuleImplApi;
+import cz.metacentrum.perun.core.implApi.modules.attributes.SkipValueCheckDuringDependencyCheck;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Module for ad DisplayName defines way how to count a final value for display name of group in o365mu from existing attributes.
+ * It expects existence of some attributes and hierarchy of groups.
+ *
+ * @author Michal Stava <stavamichal@gmail.com>
+ */
+@SkipValueCheckDuringDependencyCheck
+public class urn_perun_group_attribute_def_virt_adDisplayName_o365mu extends GroupVirtualAttributesModuleAbstract implements GroupVirtualAttributesModuleImplApi {
+
+	private final static Logger log = LoggerFactory.getLogger(urn_perun_group_attribute_def_virt_adDisplayName_o365mu.class);
+
+	private static final String A_G_D_AD_DISPLAY_NAME_O365MU = AttributesManager.NS_GROUP_ATTR_DEF + ":adDisplayName:o365mu";
+	private static final String A_G_D_INET_CISPR = AttributesManager.NS_GROUP_ATTR_DEF + ":inetCispr";
+	private static final String A_G_D_INET_GROUP_NAME_CS = AttributesManager.NS_GROUP_ATTR_DEF + ":inetGroupNameCS";
+	private static final String A_G_D_INET_GROUP_NAME_ABB_EN = AttributesManager.NS_GROUP_ATTR_DEF + ":inetGroupNameAbbEN";
+	private static final String A_G_D_INET_WORKPLACES_TYPE_CS = AttributesManager.NS_GROUP_ATTR_DEF + ":inetWorkplacesTypeCS";
+
+	private static final Pattern partOfUniversityPattern = Pattern.compile("^[1-9][1-9][0]{4}$");
+	private static final Pattern otherWorkplacesPattern = Pattern.compile("^[1-9][0-9]{5}$");
+
+	private static final String TOP_LEVEL_GROUP_CISPR = "000000";
+	private static final String TOP_LEVEL_PREFIX = "MUNI";
+
+	@Override
+	public Attribute getAttributeValue(PerunSessionImpl sess, Group group, AttributeDefinition attributeDefinition) {
+		Attribute attribute = new Attribute(attributeDefinition);
+
+		String defaultAdDisplayName = getStringValueOfGroupAttribute(sess, group, A_G_D_AD_DISPLAY_NAME_O365MU);;
+		if(defaultAdDisplayName != null) {
+			//there is a default value and we can take it
+			attribute.setValue(defaultAdDisplayName);
+			return attribute;
+		}
+
+		//If there is no way to take default value, count a new one from other attributes (if possible)
+		String cispr = getStringValueOfGroupAttribute(sess, group, A_G_D_INET_CISPR);
+		String typeOfWorkplaces = getStringValueOfGroupAttribute(sess, group, A_G_D_INET_WORKPLACES_TYPE_CS);
+		//if any of these values are null, return empty attribute instead
+		if(cispr == null || typeOfWorkplaces == null) return attribute;
```

Comment:
This is just a performance-related question, but wouldn't it be better to check the nullity of the cispr right away? This way, if the cispr is null, the other attribute is read and not used at all.

Reply:
You are right, changed.